cupy/manipulation/add_remove.py

```diff
@@ -7,10 +10,21 @@
 # TODO(okuta): Implement insert
 
 
-# TODO(okuta): Implement append
+def append(arr, values, axis=None):
+    if not isinstance(values, cupy.ndarray):
+        values = cupy.array(numpy.asarray(values))
+    if axis is None:
+        arr = arr.ravel()
+        values = values.ravel()
+        axis = 0
+    return core.concatenate_method((arr, values), axis)
 
 
-# TODO(okuta): Implement resize
+def resize(a, new_shape):
+    a = a.ravel()
+    reps, remainder = divmod(numpy.prod(new_shape), len(a))
+    return core.concatenate_method(
+        [a] * reps + [a[:remainder]], 0).reshape(new_shape)
```

Comment:
Sorry if this is a naive question, would it be possible to use `realloc` here?

Reply:
Maybe this could use `tile` instead.

```python
cupy.tile(a, reps)
```

Reply:
Actually better still one can use `as_strided`. This is nice as it views the original data (so no additional allocation) and is `O(1)` in `reps`.

```python
cupy.lib.stride_tricks.as_strided(a, (reps,) + a.shape, (0,) + a.strides).flatten()
```

Reply:
> AFAIK, CUDA does not have `realloc`.

Good point! Will take a look into this.

<hr>

> `xp.resize` need to always return a new copy anyway.

NumPy's does not always according to [their docs]( https://docs.scipy.org/doc/numpy/reference/c-api.array.html#c.PyArray_Resize ). Quoting below. Emphasis is my own.

> It changes the shape of `self` inplace and will reallocate the memory for `self` **if** `newshape` has a different total number of elements then the old shape.

Not sure we need to follow this recipe exactly, but figured it was worth observing at least.

<hr>

Anyways I'm guessing you were more referring to the `stride_tricks` suggestion. That's mainly a way to handle the repetition of the array's content without needing to allocate some intermediate arrays that will only later be discarded. As I'm guessing this function's performance will be dominated by the cost of memory allocation, am mainly thinking of how we can cut that down. Though could be wrong about this.

Reply:
> NumPy's does not always according to their docs. Quoting below. Emphasis is my own.

https://numpy.org/doc/1.17/reference/c-api.array.html#c.PyArray_Resize
It is a description of `numpy.ndarray.resize`, not `numpy.resize`, and exactly is it the feature you want?

`numpy.ndarray.resize` changes the shape in-place and will reallocate memory space if necessary, but it works only when `self` owns its data and no other arrays refer to the memory space.